virttest/qemu_vm.py

```diff
@@ -1632,83 +1534,47 @@ def sort_key(dev):
                 devices.insert(pvpanic_dev)
 
         # Add serial console redirection
-        for serial in params.objects("serials"):
+        serials = params.objects('serials')
+        if serials:
+            self.serial_session_device = serials[0]
+            host = params.get('chardev_host', '127.0.0.1')
+            free_ports = utils_misc.find_free_ports(
+                5000, 6000, len(serials), host)
+            reg_count = 0
+        self.virtio_ports = []
+        for index, serial in enumerate(serials):
             serial_filename = vm.get_serial_console_filename(serial)
-            cmd = add_serial(devices, serial, serial_filename)
-            devices.insert(StrDev('SER-%s' % serial, cmdline=cmd))
+            serial_params = params.object_params(serial)
+            # Workaround for console issue, details:
+            # http://lists.gnu.org/archive/html/qemu-ppc/2013-10/msg00129.html
+            if 'ppc' in params.get('vm_arch_name', arch.ARCH)\
+                    and serial_params.get('serial_type') == 'spapr-vty':
+                reg = 0x30000000 + 0x1000 * reg_count
+                serial_params['serial_reg'] = hex(reg)
+                reg_count += 1
+            backend = serial_params.get('chardev_backend',
+                                        'unix_socket')
+            if backend in ['udp', 'tcp_socket']:
+                serial_params['chardev_host'] = host
+                serial_params['chardev_port'] = free_ports[index]
+            serial_devices = devices.serials_define_by_params(
+                serial, serial_params, serial_filename)
+
+            devices.insert(serial_devices)
+
+            # Create virtio_ports (virtio_serialports and virtio_consoles)
+            serial_type = serial_params['serial_type']
+            if serial_type.startswith('virt'):
```

Comment:
For making code more readable, please use "virtio" instead "virt"

Reply:
@xutian  serial_type is `virtserialport` and `virtconsole`, so we need use `virt` here :)

Reply:
@xutian They are `virtserialport` and `virtconsole`, do no start with `virtio`.

Reply:
OK, please fix in your comments in your code